marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java

```diff
@@ -1,477 +1,187 @@
-    }
+    public RunFlowResponse runFlow(String flowName, List<String> stepNums) {
+        Flow flow = flowManager.getFlow(flowName);
 
-    @Override
-    public FlowRunner onItemComplete(FlowItemCompleteListener listener) {
-        this.flowItemCompleteListeners.add(listener);
-        return this;
-    }
+        //Validation of flow, provided steps
+        if (flow == null){
+            throw new RuntimeException("Flow " + flowName + " not found");
+        }
+        Iterator<String> stepItr = stepNums.iterator();
+        Queue<String> stepsQueue = new ConcurrentLinkedQueue<>();
+        while(stepItr.hasNext()) {
+            String stepNum = stepItr.next();
+            Step tmpStep = flow.getStep(stepNum);
+            if(tmpStep == null){
+                throw new RuntimeException("Step " + stepNum + " not found in the flow");
+            }
+            stepsQueue.add(stepNum);
+        }
 
-    @Override
-    public FlowRunner onItemFailed(FlowItemFailureListener listener) {
-        this.flowItemFailureListeners.add(listener);
-        return this;
-    }
+        String jobId = UUID.randomUUID().toString();
+        RunFlowResponse response = new RunFlowResponse(jobId);
 
-    @Override
-    public FlowRunner onStatusChanged(FlowStatusListener listener) {
-        this.flowStatusListeners.add(listener);
-        return this;
-    }
+        //Put response, steps and flow in maps with jobId as key
+        flowResp.put(jobId, response);
+        stepsMap.put(jobId, stepsQueue);
+        flowMap.put(jobId, flow);
 
-    @Override
-    public FlowRunner onFinished(FlowFinishedListener listener) {
-        this.flowFinishedListeners.add(listener);
-        return this;
-    }
+        //add jobId to a queue
+        jobQueue.add(jobId);
 
-    @Override
-    public void awaitCompletion() {
-        try {
-            awaitCompletion(Long.MAX_VALUE, TimeUnit.DAYS);
-        } catch (InterruptedException| TimeoutException e) {
+        if(! isRunning.get()){
+            initializeFlow(jobId);
         }
+        return response;
     }
 
-    @Override
-    public void awaitCompletion(long timeout, TimeUnit unit) throws InterruptedException,TimeoutException {
-        if (runningThread != null) {
-            runningThread.join(unit.convert(timeout, unit));
-            if (runningThread.getState() != Thread.State.TERMINATED) {
-                if ( dataMovementManager != null && queryBatcher != null ) {
-                    dataMovementManager.stopJob(queryBatcher);
-                }
-                runningThread.interrupt();
-                throw new TimeoutException("Timeout occurred after "+timeout+" "+unit.toString());
-            }
+    private void initializeFlow(String jobId) {
+        isRunning.set(true);
+        runningJobId = jobId;
+        runningFlow = flowMap.get(runningJobId);
+        if(threadPool == null || threadPool.isTerminated()) {
+            threadPool = Executors.newFixedThreadPool(1);
```

Comment:
marklogic-data-hub/marklogic-data-hub/src/main/java/com/marklogic/hub/flow/impl/FlowRunnerImpl.java#L103

Why not give more thread pool, for example, 4?

Reply:
The requirement was to run one flow at a time and queue other flows  and execute them in order when runFlow is called . So, a single thread seemed sufficient for this